File: src/2025/XPrep/153.cc
#include "bits/stdc++.h";
using namespace std;

class Solution {
  public:
    int findMin(vector<int> &nums) {
        int minVal = INT_MAX;
        int n = nums.size();
        int l = 0, r = n - 1;

        if (nums[l] < nums[r]) return nums[l];

        while (l <= r) {
            int m = l + (r - l) / 2;
            minVal = min(minVal, nums[m]);
            // something like [4,5,6,7,0,1,2]
            if (nums[m] > nums[r]) {
                // min element is on the right
                l = m + 1;
            } else {
                r = m - 1;
            }
        }

        return minVal;
    }
};

// 6 7 0 1 2 4 5

File: src/2025/rusting/153.rs
struct Solution;

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let (mut l, mut r) = (0, (nums.len() - 1) as i32);
        
        while l < r {
            let m = l + (r - l) / 2;
            let middle = nums[m as usize];

            if middle > nums[r as usize] {
                l = m + 1;
            } else {
                r = m;
            }
        }

        nums[l as usize]
    }
}